python-spec/src/somacore/query/fast_csr.py

```diff
@@ -0,0 +1,291 @@
+import concurrent.futures as futures
+import os
+from typing import List, Tuple, Type, cast
+
+import numba
+import numpy as np
+import numpy.typing as npt
+import pandas as pd
+import pyarrow as pa
+from scipy import sparse
+
+from somacore.data import SparseNDArray
+
+from .eager_iter import EagerIterator
+
+
+@numba.jit(nopython=True, nogil=True)  # type: ignore
+def _accum_row_length(row_length, row_ind):
+    for i in range(len(row_ind)):
+        row_length[row_ind[i]] += 1
+    return None
+
+
+@numba.jit(nopython=True, nogil=True)  # type: ignore
+def _copy_chunk_range(
+    data_chunk,
+    row_ind_chunk,
+    col_ind_chunk,
+    data,
+    indices,
+    indptr,
+    row_rng_mask,
+    row_rng_val,
+):
+    for n in range(len(data_chunk)):
+        row = row_ind_chunk[n]
+        if (row & row_rng_mask) != row_rng_val:
+            continue
+        ptr = indptr[row]
+        indices[ptr] = col_ind_chunk[n]
+        data[ptr] = data_chunk[n]
+        indptr[row] += 1
+    return None
+
+
+@numba.jit(nopython=True, nogil=True)  # type: ignore
+def _copy_chunklist_range(
+    chunk_list: numba.typed.List,
+    data,
+    indices,
+    indptr,
+    row_rng_mask_bits,
+    job,
+):
+    row_rng_mask = (2**64 - 1) >> row_rng_mask_bits << row_rng_mask_bits
+    row_rng_val = job << row_rng_mask_bits
+    for data_chunk, row_ind_chunk, col_ind_chunk in chunk_list:
+        _copy_chunk_range(
+            data_chunk,
+            row_ind_chunk,
+            col_ind_chunk,
+            data,
+            indices,
+            indptr,
+            row_rng_mask,
+            row_rng_val,
+        )
+    return None
+
+
+@numba.jit(nopython=True, nogil=True)  # type: ignore
+def _finalize_indptr(indptr):
+    prev = 0
+    for r in range(len(indptr)):
+        t = indptr[r]
+        indptr[r] = prev
+        prev = t
+    return None
+
+
+def _select_dtype(
+    maxval: int,
+) -> Type[np.signedinteger]:
+    """
+    Ascertain the "best" dtype for a zero-based index. Given our
+    goal of minimizing memory use, "best" is currently defined as
+    smallest.
+    """
+    if maxval > np.iinfo(np.int32).max:
+        return np.int64
+    else:
+        return np.int32
+
+
+def _reindex_and_cast(
+    index: pd.Index, ids: npt.NDArray[np.int64], target_dtype: npt.DTypeLike
+) -> npt.NDArray[np.int64]:
+    return cast(
+        npt.NDArray[np.int64], index.get_indexer(ids).astype(target_dtype, copy=False)
+    )
+
+
+class CSRAccumulator:
```

Comment:
Since this class doesn’t need the presence of all the above private functions until its own methods are called, I would put this first in the file so the most important stuff is above the fold (and can give useful context to the internal helpers defined thereafter).